### ISSUE TITLE
Fluxd becomes unresponsive if any of the specified InfluxDB instances is not available

### ISSUE THREAD

**Issue by akurilov** [influxdata/platform#171]
# Steps

1. Deploy several InfluxDB instances w/ configured IFQL
2. Deploy some metrics source, for example Telegraf instance.
3. Deploy nightly Fluxd instance using the following example command:
  `./fluxd --storage-hosts <INFLUXDB_HOST1>:8082,<INFLUXDB_HOST2>:8082,...`
4. Check that Fluxd works normally, querying it with the following example command:
  ```
curl -XPOST --data-urlencode 'q=from(db: "telegraf")   |> range(start: -10m)' http://<FLUXD_HOST>/v1/query?orgID=00
  ```
5. Stop one InfluxDB instance.
6. Repeat the step # 4
7. Start the stopped previously InfluxDB instance.
8. Repeat the step # 4
9. Restart the Fluxd
  ```
  killall fluxd
  ./fluxd --storage-hosts <INFLUXDB_HOST1>:8082,<INFLUXDB_HOST2>:8082,...
  ```
10. Repeat the step # 4

# Expected

4: Fluxd response contains the expected time series data
6: Fluxd response contains the expected time series data
8: Fluxd response contains the expected time series data
10: Fluxd response contains the expected time series data

# Observed

4: Fluxd response contains the expected time series data
6: Fluxd response is empty
8: Fluxd response is empty
10: Fluxd response contains the expected time series data

**Comment by jsternberg** [comment]
This is fixed now since we switched to using grpc.

**Comment by akurilov** [comment]
@jsternberg Does it recover when a storage host becomes back online?

**Comment by jsternberg** [comment]
Yes.

**Comment by akurilov** [comment]
Unfortunately it doesn't work in this way. The observed behaviour remains the same. Should I open new issue either you will reopen this?

**Comment by jsternberg** [comment]
I'll reopen and try it today.

**Comment by akurilov** [comment]
Updated the PR #195 in accordance with GRPC errors in case of a lost connection

**Comment by jsternberg** [comment]
Just to clarify, I'm decently certain that the multi-host functionality is not meant to serve as a high availability mechanism. @nathanielc can confirm, but the expected output should be:

1. Succeeds.
2. Fails.
3. Succeeds.
4. Succeeds.

Does this happen for you?

**Comment by pauldix** [comment]
@jsternberg multi-host should be useful either as an HA mechanism or as something for doing federated queries. The former would require deduplication of data points with the same timestamp and the latter would require a union. We'll need to do the same thing with the Prometheus data source.

**Comment by jsternberg** [comment]
@pauldix it sounds to me like we need to modify the open source version to only read from one of the queries and to do a kind of round-robin for the data. In the open source version, it's expected that all of the servers would have the same data, correct?

Although I'm not sure how useful this is. What's the end goal for how we are expecting people to run this from open source?

**Comment by akurilov** [comment]
@jsternberg 
The issue doesn't seem to regard to HA. This is a bug. First, fluxd drops all results from all the specified storage hosts if any is not online. The resulting response is empty despite of the storage hosts remaining online. Second, the connection to a storage host is not re-established if was once shutdown.

**Comment by jsternberg** [comment]
Adding this to our backlog so we can discuss this as a team.

My main concern is that I don't think it's been defined exactly how the `--storage-hosts` flag works or what it's supposed to do and how flux itself deals with missing servers in a replicated environment. This question wasn't relevant to the open source influxdb because it was a single server and never had missing data because of that. The enterprise version of that, if you ever used it, would coordinate to figure out which servers it could talk to and, if one of the servers was missing, would just return the partial data. But it would never query multiple servers with the same data.

So I'm not sure what setting multiple `--storage-hosts` means for open source in this context. For the enterprise version of this new software, I don't think we've discussed yet if we believe the 1.x behavior was correct or if it should return an error or some kind of warning when the data is incomplete. It didn't do either of those in 1.x. For the open source, what does setting multiple storage hosts mean? Is it meant as an HA mechanism or is it meant as something to retrieve data from multiple sources and return it as kind of an open source sharding mechanism so that people can build their own sharding mechanism around it?

That's why I'm not sure what the correct behavior is and why I'm leaning towards the current behavior being correct and I'm actually leaning more towards believing that multiple sources doesn't make sense for the open source product. And while it accepts multiple comma-separated parameters, it doesn't actually document exactly what that means in the current product and I'm not sure we really thought about what multiple servers meant.

Put simply, for open source, do we assume that if multiple servers are given, the servers are:

1. Identical.
2. Sharded.

We have to assume one of them or we have to add a parameter to specify the configuration so that we can properly query them. If they're identical, we query only one of them and you get a simple HA solution (that isn't as good as the enterprise, but we've already said we're not doing that in the open source). If they're sharded, you have to query all of them. Right now, I think it assumes sharded.

Sorry that was a bit wordy.

@akurilov also, I'm not sure what version you are using, but the grpc version re-establishes connections. You have to use the nightlies or build from master to get that functionality. That's the reason I initially closed this. Are you using the nightlies or building from master for both influxd and fluxd?

**Comment by akurilov** [comment]
@jsternberg 
I'm using the version from the master branch. gRPC doesn't seem to re-establish the connections, possibly due to DialOptions or so.

**Comment by pauldix** [comment]
@jsternberg it isn't expected that all the servers necessarily have the same data. I'm thinking it would either operate on the assumption that the servers have the same data or that they don't. Maybe like a flag that says if it should select one of the hosts, or send the query to all hosts and combine the results.
